bundles/core/org.eclipse.smarthome.core.scheduler/src/main/java/org/eclipse/smarthome/core/scheduler/common/ThreadPoolManager.java

```diff
@@ -217,6 +260,108 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
         }
     }
 
+    public static class ExpressionThreadPoolExecutor extends ScheduledThreadPoolExecutor {
+
+        private List<Runnable> running = Collections.synchronizedList(new ArrayList<Runnable>());
+        private Map<Runnable, Runnable> scheduled = Collections.synchronizedMap(new HashMap<Runnable, Runnable>());
+        private Map<Runnable, Future<?>> futures = Collections.synchronizedMap(new HashMap<Runnable, Future<?>>());
+
+        public ExpressionThreadPoolExecutor(final String poolName, int corePoolSize) {
+            this(poolName, corePoolSize, new NamedThreadFactory(poolName), new ThreadPoolExecutor.DiscardPolicy() {
+                // The pool is bounded and rejections will happen during shutdown
+                @Override
+                public void rejectedExecution(Runnable runnable, ThreadPoolExecutor threadPoolExecutor) {
+                    // Log and discard
+                    logger.warn("Thread pool '{}' rejected execution of {}",
+                            new Object[] { poolName, runnable.getClass() });
+                    super.rejectedExecution(runnable, threadPoolExecutor);
+                }
+            });
+        }
+
+        public ExpressionThreadPoolExecutor(String threadPool, int corePoolSize, ThreadFactory threadFactory,
+                RejectedExecutionHandler rejectedHandler) {
+            super(corePoolSize, threadFactory, rejectedHandler);
+        }
+
+        @Override
+        protected void beforeExecute(Thread thread, Runnable runnable) {
+            super.beforeExecute(thread, runnable);
+            running.add(runnable);
+        }
+
+        @Override
+        protected void afterExecute(Runnable runnable, Throwable throwable) {
+            super.afterExecute(runnable, throwable);
+            running.remove(runnable);
+            futures.remove(runnable);
+            scheduled.remove(runnable);
+            if (throwable != null) {
+                Throwable cause = throwable.getCause();
+                if (cause instanceof InterruptedException) {
+                    // Ignore this, might happen when we shutdownNow() the executor. We can't
+                    // log at this point as the logging system might be stopped already.
+                    return;
+                }
+            }
+        }
+
+        public void schedule(final Runnable task, final Expression expression) {
+            if (task == null) {
+                throw new NullPointerException();
+            }
+
+            this.setCorePoolSize(this.getCorePoolSize() + 1);
```

Comment:
You are too quick here I think. Expressions  can have multiple execution moments, so, in order to keep it simple there is one thread managing each expression, e.g. the scheduleTask.Putting the thread to sleep will not hurt the system either I think. IMHO things will get complicated if you want to put that in a single thread. Personally, I would refrain from interfering with the inner bodyworks of ScheduledThreadPoolExecutor (delayedExecute is private). If you put it in 1 thread, that thread would have to run like crazy in order to constantly evaluate the expression execution moments, no? In the best case you would have to evaluate all getTimeAfter(now) of all Expressions, and put it to sleep to the smallest time that comes up; and then you have to manage that thread every time a new expression is schedule()ed. 

On the pool size you might be right, but I don't remember if ScheduledThreadPoolExecutor automatically reduces that after a while.